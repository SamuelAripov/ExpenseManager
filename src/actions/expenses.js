import database from '../firebase/firebase';
import { get } from 'http';

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData
        const expense = { description, note, amount, createdAt }

        return database.ref(`users/${uid}/expenses`).push(expense).then(({ ref }) => {
            console.log(ref)
            dispatch(addExpense({
                id: expense.key,
                ...expense
            }))
        })
    }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
    type: 'REMOVE_EXPENSE',
    id
})
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})
export const clearExpenses = () => ({
    type: 'CLEAR_EXPENSES'
})
export const startSetExpense = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        dispatch(clearExpenses())
        return database
            .ref(`users/${uid}/expenses/`)
            .once('value')
            .then((snapshot) => {
                snapshot.forEach((childSnapshot) => {
                    dispatch(addExpense({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    }))
                    console.log('Official Data', {
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    })
                })
            }).catch(e => console.log(e))
    }
}
export const startRemoveExpense = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({ id }))
            })
    }
}
export const startEditExpense = (id, update) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .update({
                ...update
            })
            .then(() => {
                dispatch(editExpense(id, update))
            })
    }
}

