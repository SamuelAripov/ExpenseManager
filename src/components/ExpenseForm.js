import React from 'react'
import DatePicker from 'react-datepicker'
export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props)
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? props.expense.amount : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? props.expense.createdAt : new Date().valueOf(),
            calendarFocused: false
        }
    }
    
    onDescripitionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onTextFormChange = (e) => {
        const textValue = e.target.value
        this.setState(() => ({ note: textValue }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }))
    }
    onSubmit = (e) => {
        e.preventDefault()
        this.props.onSubmit({
            description: this.state.description,
            amount: this.state.amount,
            createdAt: this.state.createdAt,
            note: this.state.note
        })
    }
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescripitionChange}
                        required
                    />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        required
                    />
                    <br />
                    <DatePicker
                        selected={this.state.createdAt}
                        onChange={this.onDateChange}
                        dateFormat="MMMM d, yyyy"
                        onChangeRaw={(e) => e.preventDefault()}
                    />
                    <textarea
                        placeholder="Add a note for you expense (optional)"
                        value={this.state.note}
                        onChange={this.onTextFormChange}
                    >
                    </textarea>
                    <button>Submit</button>
                </form>
            </div>
        )
    }
}