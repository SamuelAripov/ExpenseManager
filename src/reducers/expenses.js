// Expenses Reducer
import database from '../firebase/firebase'

export default (state = [], action) => {
    switch (action.type) {
        case 'ADD_EXPENSE': {
            console.log(action.expense)
            return [
                ...state,
                action.expense
            ]
        }
        case 'REMOVE_EXPENSE': {
            return state.filter((expense) => expense.id !== action.id)
        }
        case 'EDIT_EXPENSE': {
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else
                    return expense
            })
        }
        case 'SET_EXPENSES': {
            return action.expenses
        }
        case 'CLEAR_EXPENSES': {
            return []
        }
        default:
            return state;
    }
}