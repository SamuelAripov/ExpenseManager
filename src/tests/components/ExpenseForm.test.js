import React from 'react'
import { shallow } from 'enzyme'
import ExpenseForm from '../../components/ExpenseForm'
import expenses from '../fixtures/expenses'

test('should render expense form correctly', () => {
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render expense form with expense data', () => {
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} />)
    expect(wrapper).toMatchSnapshot()
})

test('should set description on input change', () => {
    const value = 'description'
    const wrapper = shallow(<ExpenseForm />)
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    })
    expect(wrapper.state('description')).toBe(value)
})

test('should set note on textarea change', () => {
    const value = 'note'
    const wrapper = shallow(<ExpenseForm />)
    wrapper.find('textarea').at(0).simulate('change', {
        target: { value }
    })
    expect(wrapper.state('note')).toBe(value)
})

test('should set amount to a number 122.500 and get ERROR', () => {
    const value = '122.500'
    const wrapper = shallow(<ExpenseForm />)
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    })
    expect(wrapper.state('amount')).toBe("")
})

test('should set amount to a number 24.99', () => {
    const value = '24.99'
    const wrapper = shallow(<ExpenseForm />)
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    })
    expect(wrapper.state('amount')).toBe(value)
})

test('should call on submit call for valid form submition', () => {
    const onSubmitSpy = jest.fn()
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />)
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    })
    expect(onSubmitSpy).toHaveBeenLastCalledWith({ "amount": 3, "createdAt": 1, "description": "1", "note": "" })
})

test('should set new date on date change', () => {
    const now = new Date().valueOf()
    const wrapper = shallow(<ExpenseForm />)
    wrapper.find('DatePicker').simulate('submit', {
        target: { now }
    })
    expect(wrapper.state('createdAt')).toEqual(now)
})