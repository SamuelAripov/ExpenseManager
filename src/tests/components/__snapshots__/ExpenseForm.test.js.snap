// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense form correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      required={true}
      type="text"
      value="1"
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      required={true}
      type="number"
      value={3}
    />
    <br />
    <DatePicker
      allowSameDay={false}
      dateFormat="MMMM d, yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onChangeRaw={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for you expense (optional)"
      value=""
    />
    <button>
      Submit
    </button>
  </form>
</div>
`;

exports[`should render expense form with expense data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      required={true}
      type="text"
      value="1"
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      required={true}
      type="number"
      value={3}
    />
    <br />
    <DatePicker
      allowSameDay={false}
      dateFormat="MMMM d, yyyy"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      nextMonthButtonLabel="Next month"
      onBlur={[Function]}
      onChange={[Function]}
      onChangeRaw={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      readOnly={false}
      renderDayContents={[Function]}
      selected={1}
      shouldCloseOnSelect={true}
      showMonthYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for you expense (optional)"
      value=""
    />
    <button>
      Submit
    </button>
  </form>
</div>
`;
